Source/UI/NewUI/PaymentForm/View/SecurityCodeView.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  SecurityCodeView.swift
+//  
+//
+//  Created by Ehab Alsharkawy
+//  Copyright © 2022 Checkout. All rights reserved.
+
+import UIKit
+import Checkout
+
+public final class SecurityCodeView: UIView {
+
+  private let securityCodeCount = 4
+  private var cardValidator: CardValidator
+  private(set) var style: CellTextFieldStyle?
+
+  private lazy var codeInputView: InputView = {
+    let view = InputView().disabledAutoresizingIntoConstraints()
+    view.delegate = self
+    return view
+  }()
+
+  init(cardValidator: CardValidator) {
+    self.cardValidator = cardValidator
+    super.init(frame: .zero)
+
+    //setup security code view
+    addSubview(codeInputView)
+    codeInputView.setupConstraintEqualTo(view: self)
+  }
+
+  required init?(coder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  func update(style: CellTextFieldStyle?) {
+    self.style = style
+    self.style?.textfield.isSupportingNumericKeyboard = true
+    codeInputView.update(style: self.style)
+  }
+
+  private func updateErrorViewStyle(isHidden: Bool, textfieldText: String?){
+    style?.error?.isHidden = isHidden
+    style?.textfield.text = textfieldText ?? ""
+    codeInputView.update(style: style)
+  }
+}
+
+extension SecurityCodeView: TextFieldViewDelegate {
+  func textFieldShouldBeginEditing(textField: UITextField) {}
+  func textFieldShouldReturn() -> Bool {  return false }
+  func textFieldShouldEndEditing(textField: UITextField, replacementString: String) -> Bool {
+    if textField.text?.count ?? 0 < securityCodeCount  {
+      updateErrorViewStyle(isHidden: false, textfieldText: textField.text)
+    }
+    return true
+  }
+
+  func textFieldShouldChangeCharactersIn(textField: UITextField, replacementString string: String) {
+    codeInputView.textFieldContainer.layer.borderColor = style?.textfield.focusBorderColor.cgColor
+  }
+
+  func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+    updateErrorViewStyle(isHidden: true, textfieldText: textField.text)
+    guard range.location < securityCodeCount else { return false }
+    guard range.length == 0 else { return true }
+    //check for max length including added spacers which all equal to 5
+    guard !string.isEmpty else { return false }
+    let replacementText = string.replacingOccurrences(of: " ", with: "")
```

Comment:
I have no idea why we check if length is 0, why there's a comment after it and why removing text is not allowed... IMO this method doesn't read clear at all.

Reply:
range length should be 0 every time

Reply:
Tried this hands on, and I don't think spaces should be allowed at all. I tried breaking down my understanding of this code, and I think most is not helpful.
```swift
// Prevent input longer then logical limit of security code
guard range.location < maxSecurityCodeLenght else { return false }

// Standard input will not replace characters, so range will be zero
// !!! You might want to remove this bit as if a paste occurs over existing input, the next validation will be returned, as you are then replacing existing content with new one and range will not be 0 !!!
guard range.lenght == 0 else { return true }

// I am not sure why you need this. An input of "" wouldn't really be possible anyway?
guard !string.isEmpty else { return false }

// you intend to only have integers, so why would we want to accept spaces?
let replacementText = string.replacingOccurrences(of: " ", with: "")
```

Tried this code and it seems to work the same, do you think this would work from your view?
```swift
// Prevent input longer then logical limit of security code
guard range.location < maxSecurityCodeLenght else { return false }
// Enable deleting of text
if string.isEmpty { return true }
// Prevent non numeric input from being inserted
return Int(string) != nil
```